core/drivers/stm32_i2c.c

```diff
@@ -0,0 +1,1383 @@
+	uintptr_t base = 0;
+	uint32_t val = 0;
+
+	hi2c->base.pa = init_data->pbase;
+	hi2c->clock = init_data->clock;
+
+	rc = i2c_setup_timing(hi2c, init_data, &timing);
+	if (rc)
+		return rc;
+
+	stm32_clock_enable(hi2c->clock);
+	base = get_base(hi2c);
+	hi2c->i2c_state = I2C_STATE_BUSY;
+
+	/* Disable the selected I2C peripheral */
+	io_clrbits32(base + I2C_CR1, I2C_CR1_PE);
+
+	/* Configure I2Cx: Frequency range */
+	write32(timing & TIMINGR_CLEAR_MASK, base + I2C_TIMINGR);
+
+	/* Disable Own Address1 before set the Own Address1 configuration */
+	write32(0, base + I2C_OAR1);
+
+	/* Configure I2Cx: Own Address1 and ack own address1 mode */
+	if (init_data->addr_mode_10b_not_7b)
+		write32(I2C_OAR1_OA1EN | I2C_OAR1_OA1MODE |
+			init_data->own_address1,
+			base + I2C_OAR1);
+	else
+		write32(I2C_OAR1_OA1EN | init_data->own_address1,
+			base + I2C_OAR1);
+
+	/* Configure I2Cx: Addressing Master mode */
+	write32(0, base + I2C_CR2);
+	if (init_data->addr_mode_10b_not_7b)
+		io_setbits32(base + I2C_CR2, I2C_CR2_ADD10);
+
+	/*
+	 * Enable the AUTOEND by default, and enable NACK
+	 * (should be disabled only during Slave process).
+	 */
+	io_setbits32(base + I2C_CR2, I2C_CR2_AUTOEND | I2C_CR2_NACK);
+
+	/* Disable Own Address2 before set the Own Address2 configuration */
+	write32(0, base + I2C_OAR2);
+
+	/* Configure I2Cx: Dual mode and Own Address2 */
+	if (init_data->dual_address_mode)
+		write32(I2C_OAR2_OA2EN | init_data->own_address2 |
+			(init_data->own_address2_masks << 8),
+			base + I2C_OAR2);
+
+	/* Configure I2Cx: Generalcall and NoStretch mode */
+	val = 0;
+	if (init_data->general_call_mode)
+		val |= I2C_CR1_GCEN;
+	if (init_data->no_stretch_mode)
+		val |= I2C_CR1_NOSTRETCH;
+	write32(val, base + I2C_CR1);
+
+	/* Enable the selected I2C peripheral */
+	io_setbits32(base + I2C_CR1, I2C_CR1_PE);
+
+	hi2c->i2c_err = I2C_ERROR_NONE;
+	hi2c->i2c_state = I2C_STATE_READY;
+
+	rc = i2c_config_analog_filter(hi2c, init_data->analog_filter);
+	if (rc)
+		EMSG("I2C analog filter error %d", rc);
+
+	stm32_clock_disable(hi2c->clock);
+
+	return rc;
+}
+
+/* I2C transmit (TX) data register flush sequence */
+static void i2c_flush_txdr(struct i2c_handle_s *hi2c)
+{
+	uintptr_t base = get_base(hi2c);
+
+	/*
+	 * If a pending TXIS flag is set,
+	 * write a dummy data in TXDR to clear it.
+	 */
+	if (read32(base + I2C_ISR) & I2C_ISR_TXIS)
+		write32(0, base + I2C_TXDR);
+
+	/* Flush TX register if not empty */
+	if ((read32(base + I2C_ISR) & I2C_ISR_TXE) == 0)
+		io_setbits32(base + I2C_ISR, I2C_ISR_TXE);
+}
+
+/*
+ * Wait for one or more I2C_ISR bits reach their target value
+ *
+ * @hi2c: I2C handle structure
+ * @value_mask: Bit mask for the bit positions to consider in @awaited_value
+ * @awaited_value: The awaited value of I2C_ISR once masked with @value_mask
+ * @timeout_ref: Expriation timeout reference
+ * Return 0 on success or a negative value.
+ */
+static int wait_isr_bits(struct i2c_handle_s *hi2c, uint32_t value_mask,
```

Comment:
From an internal discussion, it appears that the description here is false and use of this function is confusing. I'll push an update to clarify.